### PR TITLE
Migrate template to Gradle 6.7 and AGP 4.1.0

### LINKED ISSUE #49438: Upgrade the Gradle version in the project template
Gradle 6.0 came out recently: https://docs.gradle.org/6.0/release-notes.html

The Flutter template currently uses 5.6. However, 6.0 might become the default version for new Android apps generated by Android Studio.

For context, this is already the case: https://github.com/flutter/flutter/pull/46260#issuecomment-562802455 if a Flutter app is edited on Android Studio.

### LINKED ISSUE #58247: flutter build apk fails on an app using Android Gradle plugin 4.0
# Canonical Issue
https://issuetracker.google.com/issues/141126614
----------------

## Steps to Reproduce
1. Run `flutter create counter`
2. Update the app to use Android Gradle plugin 4.0.0
3. Run `flutter build apk`

## Logs
```
FAILURE: Build failed with an exception.                                
                                                                        
* What went wrong:                                                      
Execution failed for task ':app:lintVitalRelease'.                      
> Could not resolve all artifacts for configuration ':app:debugRuntimeClasspath'.
   > Failed to transform libs.jar to match attributes {artifactType=processed-jar, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for JetifyTransform: /Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar.
         > Failed to transform '/Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar' using Jetifier. Reason: FileNotFoundException, message: /Users/michaelbui/Projects/counter/build/app/intermediates/flutter/debug/libs.jar (No such file or directory). (Run with --stacktrace for more details.)
           Please file a bug at http://issuetracker.google.com/issues/new?component=460323.
                                                                        
* Try:                                                                  
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
                                                                        
* Get more help at https://help.gradle.org
```

```
[✓] Flutter (Channel stable, v1.17.2, on Mac OS X 10.15.5 19F96, locale en-AU)
    • Flutter version 1.17.2 at /Users/michaelbui/flutter
    • Framework revision 5f21edf8b6 (20 hours ago), 2020-05-28 12:44:12 -0700
    • Engine revision b851c71829
    • Dart version 2.8.3

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/michaelbui/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.45.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.2
```

Edit: created an issue against the Jetifier component as suggested by Gradle output https://issuetracker.google.com/u/1/issues/157728192

### PR DESCRIPTION
**Pull request by blasten**
## Description

Makes the plugin and the tool compatible with the latest release of AGP version, and Gradle.
Migrates the template files for apps, plugins and modules.
Migrates files in the Flutter repo.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/58247
Fixes https://github.com/flutter/flutter/issues/49438

* https://issuetracker.google.com/issues/158060799
* https://issuetracker.google.com/issues/158753935

## Tests

I added the following tests: unit & e2e tests

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Zazo032** [comment]
@blasten latest available versions right now are `6.7.1` for Gradle and `4.1.1` for AGP, just in case you want to update it to latest available patch

File: dev/benchmarks/complex_layout/android/app/build.gradle [hunk 1]
```diff
@@ -19,15 +19,11 @@ apply plugin: 'com.android.application'
 apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    compileSdkVersion 29
-
-    lintOptions {
-        disable 'InvalidPackage'
-    }
+    compileSdkVersion 30
```
Comment (xster) [inline dev/benchmarks/complex_layout/android/app/build.gradle:22]: this is ok but just for my understanding, this is orthogonal to a agp move right?

File: dev/benchmarks/complex_layout/android/gradle.properties [hunk 2]
```diff
@@ -1,4 +1,3 @@
 org.gradle.jvmargs=-Xmx1536M
 android.useAndroidX=true
 android.enableJetifier=true
-android.enableR8=true
```
Comment (xster) [inline dev/benchmarks/complex_layout/android/gradle.properties:4]: can you elaborate? Or is this just to save time since no one is shipping it?

File: dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java [hunk 3]
```diff
@@ -7,20 +7,12 @@
...
-import io.flutter.plugin.common.MethodCall;
-import io.flutter.plugin.common.MethodChannel;
 import io.flutter.plugins.GeneratedPluginRegistrant;
 
 public class MainActivity extends FlutterActivity {
   @Override
   protected void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
     GeneratedPluginRegistrant.registerWith(this);
-    new MethodChannel(getFlutterView(), "flavor").setMethodCallHandler(new MethodChannel.MethodCallHandler() {
```
Comment (xster) [inline dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java:19]: related?

File: packages/flutter_tools/gradle/flutter.gradle [hunk 4]
```diff
@@ -817,8 +840,56 @@ class FlutterPlugin implements Plugin<Project> {
...
+                    // This mapping is based on the following rules:
+                    // 1. If the host app build variant name is `profile` then the equivalent
+                    //    Flutter variant is `profile`.
+                    // 2. If the host app build variant is debuggable
+                    //    (e.g. `buildType.debuggable = true`), then the equivalent Flutter
+                    //    variant is `debug`.
+                    // 3. Otherwise, the equivalent Flutter variant is `release`.
+                    String variantBuildMode = buildModeFor(libraryVariant.buildType)
+                    if (buildModeFor(appProjectVariant.buildType) != variantBuildMode) {
+                        return
```
Comment (xster) [inline packages/flutter_tools/gradle/flutter.gradle:881]: Can we print a message here? I always found the mismatch error a bit cryptic.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 5]
```diff
@@ -588,14 +588,9 @@ abstract class FlutterCommand extends Command<void> {
...
-      defaultsTo: true,
-      help: 'Whether to enable code shrinking on release mode. '
-            'When enabling shrinking, you also benefit from obfuscation, '
-            'which shortens the names of your app’s classes and members, '
-            'and optimization, which applies more aggressive strategies to '
-            'further reduce the size of your app. '
-            'To learn more, see: https://developer.android.com/studio/build/shrink-code',
-      );
+      hide: true,
+      help: 'This flag is deprecated.'
```
Comment (xster) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:592]: Describe a bit more what now happens

File: packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties [hunk 6]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-all.zip
```
Comment (xster) [inline packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties:6]: can we add another test to make sure the whole thing builds with gradle 5 / agp 3?

**Review by xster — COMMENTED** [review]
woohoo! Amazing

**Comment by xster** [comment]
@csells could be worth calling this out on the next stable blog

File: packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (blasten) [reply packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties:6]: that is already the case in the customer tests

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 5 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:592]: Done.

File: dev/benchmarks/complex_layout/android/gradle.properties (hunk 2 above)
Comment (blasten) [reply dev/benchmarks/complex_layout/android/gradle.properties:4]: R8 has become the default in this AGP version. Having this causes a warning.

File: dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java (hunk 3 above)
Comment (blasten) [reply dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java:19]: it's a small clean up. This isn't used by the test, and was causing some warnings.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 7]
```diff
@@ -158,24 +159,23 @@ class FlutterPlugin implements Plugin<Project> {
...
-                    minifyEnabled true
-                    // Enables resource shrinking, which is performed by the
-                    // Android Gradle plugin.
-                    // NOTE: The resource shrinker can't be used for libraries.
-                    shrinkResources isBuiltAsApp(project)
-                    // Fallback to `android/app/proguard-rules.pro`.
-                    // This way, custom Proguard rules can be configured as needed.
-                    proguardFiles project.android.getDefaultProguardFile("proguard-android.txt"), flutterProguardRules, "proguard-rules.pro"
-                }
+        String flutterProguardRules = Paths.get(flutterRoot.absolutePath, "packages", "flutter_tools",
```
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:162]: This is one is about removing the need to make this conditional via a Flutter tool flag. That isn't needed.  Users shouldn't disable this in release, but if they really want they can always do it in build.gradle.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 8]
```diff
@@ -158,24 +159,23 @@ class FlutterPlugin implements Plugin<Project> {
...
+                "gradle", "flutter_proguard_rules.pro")
+        project.android.buildTypes {
+            release {
```
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:165]: levers?

File: packages/flutter_tools/gradle/flutter.gradle [hunk 9]
```diff
@@ -570,20 +559,84 @@ class FlutterPlugin implements Plugin<Project> {
...
         if (target == null) {
             target = 'lib/main.dart'
         }
         if (project.hasProperty('target')) {
             target = project.property('target')
         }
+        return target
+    }
+
+    // In AGP 4.0, the Android linter task depends on all the JAR tasks containing `libapp.so`.
```
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle]: I changed the sentence a bit. https://github.com/flutter/flutter/issues/58247#issuecomment-636253593

File: packages/flutter_tools/gradle/flutter.gradle (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:882]: This is during configuration phase. It could be quite spammy, and no very meaningful for debugging. What's the user story?

File: dev/benchmarks/complex_layout/android/app/build.gradle (hunk 1 above)
Comment (blasten) [reply dev/benchmarks/complex_layout/android/app/build.gradle:22]: that's true

File: dev/benchmarks/complex_layout/android/gradle.properties (hunk 2 above)
Comment (xster) [reply dev/benchmarks/complex_layout/android/gradle.properties:4]: got it, thanks

File: dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java (hunk 3 above)
Comment (xster) [reply dev/integration_tests/gradle_deprecated_settings/android/app/src/main/java/com/yourcompany/flavors/MainActivity.java:19]: ok

File: packages/flutter_tools/gradle/flutter.gradle (hunk 8 above)
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter_tools/gradle/flutter.gradle:165]: when users create custom types, we should document what the right approach is to have the correct default settings associated. e.g. recommend

```groovy
production {
     initWith release         
}
```

rather than making one from scratch

File: packages/flutter_tools/gradle/flutter.gradle (hunk 4 above)
Comment (xster) [reply packages/flutter_tools/gradle/flutter.gradle:885]: Maybe I'm thinking of something else. In Android Studio, open build variants tab in case of gradle subproject integration, pick one  variant for the app and another variant for the flutter module.

File: packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (xster) [reply packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties:6]: which customer tests are you referring to?

File: packages/flutter_tools/gradle/flutter.gradle (hunk 8 above)
[the remark this answers is not in the document]
Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:165]: I don't think we are doing anything special, other than profile.

`release` is an official build type provided by the Android plugin. I think the Flutter docs can link to https://developer.android.com/studio/build/build-variants.

File: packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties (hunk 6 above)
Comment (blasten) [reply packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties:6]: https://github.com/flutter/tests/tree/master/registry - Although, it looks like we don't build apk in any of those. I will add one that does it.

Comment (blasten) [reply packages/integration_test/example/android/gradle/wrapper/gradle-wrapper.properties:6]: Filed https://github.com/flutter/flutter/issues/70964. Gradle 5.* is currently tested by building the Flutter Gallery demo app.

**Review by xster — APPROVED** [review]
LGTM 🤞

**Comment by amirh** [comment]
This broke flutter/plugins, see: https://cirrus-ci.com/task/6538283054268416

**Comment by gustav3d** [comment]
So instead of fiximg the boken plugin https://cirrus-ci.com/task/6538283054268416, you revert this?!

**Comment by amirh** [comment]
@gustav3d per https://github.com/flutter/flutter/wiki/Tree-hygiene if a commit breaks a tree it should be first reverted, no questions asked, then it should be re-landed with the issue figured.

**Comment by gustav3d** [comment]
Ah, ok :)
